File: server/apps/nexotype/schemas/asset/small_molecule_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class SmallMoleculeCreate(BaseModel):
    """Schema for creating a new small molecule"""
    uid: str
    name: str
    project_code: str | None = None
    smiles: str
    inchi_key: str | None = None


class SmallMoleculeUpdate(BaseModel):
    """Schema for updating a small molecule"""
    uid: str = None
    name: str = None
    project_code: str | None = None
    smiles: str = None
    inchi_key: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class SmallMoleculeDetail(BaseModel):
    """Schema for small molecule details"""
    id: int
    uid: str
    name: str
    project_code: str | None = None
    asset_type: str
    smiles: str
    inchi_key: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class SmallMoleculeListResponse(BaseModel):
    """Response schema for listing small molecules"""
    success: bool
    data: list[SmallMoleculeDetail] | None = None
    count: int = 0
    error: str | None = None


class SmallMoleculeResponse(BaseModel):
    """Response schema for single small molecule operations"""
    success: bool
    data: SmallMoleculeDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/asset/small-molecule.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const SmallMoleculeSchema = z.object({
  id: z.number(),
  uid: z.string().min(1).max(100),
  name: z.string().min(1).max(255),
  project_code: z.string().max(50).nullable().optional(),
  asset_type: z.string().max(50),
  smiles: z.string().min(1),
  inchi_key: z.string().max(27).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateSmallMoleculeSchema = z.object({
  uid: z.string().min(1).max(100),
  name: z.string().min(1).max(255),
  project_code: z.string().max(50).nullable().optional(),
  smiles: z.string().min(1),
  inchi_key: z.string().max(27).nullable().optional(),
});

export const UpdateSmallMoleculeSchema = z.object({
  uid: z.string().min(1).max(100).optional(),
  name: z.string().min(1).max(255).optional(),
  project_code: z.string().max(50).nullable().optional(),
  smiles: z.string().min(1).optional(),
  inchi_key: z.string().max(27).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type SmallMolecule = z.infer<typeof SmallMoleculeSchema>;
export type CreateSmallMolecule = z.infer<typeof CreateSmallMoleculeSchema>;
export type UpdateSmallMolecule = z.infer<typeof UpdateSmallMoleculeSchema>;

// ==========================================
// Response Types
// ==========================================

export type SmallMoleculeResponse = {
  success: boolean;
  data?: SmallMolecule;
  error?: string;
};

export type SmallMoleculesResponse = {
  success: boolean;
  data?: SmallMolecule[];
  count?: number;
  error?: string;
};
